[WooPOS] [Barcodes] Support Barcode Scanning when VoiceOver is enabled

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

More context: pdfdoF-7Nl-p2

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Barcode scanner input is not received through `GameController` framework `keyChangeHandler` if VoiceOver is enabled, which looks to be a system limitation.

As a lightweight workaround, when VoiceOver is enabled, get input through UIViewController.pressesBegan -> Map UIKey to GCKeyCode -> Continue using the same scanning infrastructure with GameControllerBarcodeObserver and GameControllerBarcodeParser.  This reintroduces some of the code that was removed in the https://github.com/woocommerce/woocommerce-ios/pull/15877 PR.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Enable VoiceOver
2. Open POS
3. Confirm that barcode scanning in the item list and the setup flow continues to work and items are correctly scanned

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Tested on iPad Air iOS26 with Inateck scanner:
- Set up flow can be completed with VoiceOver enabled
- Scanning various types of barcodes in the item list works
- Scanning is not working when modals/sheets are presented with VoiceOver disabled/enabled
- Scanning starts a new order
- Tracking, including device connection and disconnection during the setup flow, works

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
VoiceOver support for Barcode Scanning

Barcode scanner  input is not received through GameController framework keyChangeHandler if VoiceOver is enabled, which looks to be a system limitation.

As a lightweight workaround, when VoiceOver is enabled, get input through UIViewController.pressesBegan -> Map UIKey to GCKeyCode -> Continue using the same scanning infrastructure with GameControllerBarcodeObserver and GameControllerBarcodeParser

staskus committed:
Update code comment with clarification

staskus committed:
Set an accessibility label to barcode image

joshheald commented on WooCommerce/Classes/POS/Presentation/Barcode Scanning/BarcodeScannerContainer.swift:
```diff
@@ -70,4 +70,50 @@ final class GameControllerBarcodeScannerHostingController: UIHostingController<E
+    }
+
+    override func pressesBegan(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
+        // Don't call super to prevent system from hiding software keyboard
```
We should probably call super if VoiceOver isn't running. Perhaps it won't matter if we didn't take first responder...

staskus replied:
I tested it a few times, at least on my device, if I call super, even if VoiceOver is disabled, it prevents the software keyboard from showing up. Therefore, I left it at that.

joshheald commented on WooCommerce/Classes/POS/Presentation/Barcode Scanning/GameControllerBarcodeObserver.swift:
```diff
@@ -142,4 +143,100 @@ final class GameControllerBarcodeObserver {
         }
     }
+
+    // MARK: - VoiceOver Support
```
What do you think about putting much of this code in a different class? I think it's a little confusing, even with the comments, to have non-GameController approaches mixed into, and using, classes with GameController in the name.

I think we would want to keep the `GameControllerBarcodeObserver` unaware of this fallback, and switch between them from the `BarcodeScannerContainer` instead.

Apologies if I'm missing any reason that this is particularly tricky to do...

staskus replied:
I think that's the right idea. I chose it for simplicity since `GameControllerBarcodeObserver` has some code that I would like to reuse between the two but maybe I could move it to the container 🤔 I'll see what I can do.

staskus replied:
- I split the logic between `GameControllerBarcodeObserver` and `UIKitBarcodeObserver`
- Added VoiceOver switch in the `BarcodeScannerContainer` and unit tested it
- Moved shared tracking logic into a `BarcodeAnalyticsTracker`

joshheald reviewed (approved):
Works as expected, and using a fallback to the old code was exactly what I was going to suggest when I read the P2 earlier.

I've made a suggestion inline about structuring the code differently – let me know if you don't think we should go that route, but I think it might be clearer if we do, and potentially easier to test.

It would also be nice to have unit tests for the switch between the different paths, but it's probably quite difficult to really effectively mock the `isVoiceOverEnabled` static var, so not a must have.

staskus committed:
Create UIKitBarcodeObserver to contain UIKit-driven barcode observing logic

Move away UIKit-related barcode scanning logic, that is used when VoiceOver is enabled, away from GameControllerBarcodeObserver to UIKitBarcodeObserver

staskus committed:
Update BarcodeScannerContainer to switch between UIKit and GameController barcode observers

staskus committed:
Clean up GameControllerBarcodeParserTests

staskus committed:
Add additional comments, make sure software keyboard is not hidden when text field is selected

pressesBegan should continue not calling super to prevent system from hiding software keyboard

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Remove UIKit import from GameControllerBarcodeObserver
